impl/src/test/java/io/kroxylicious/testing/kafka/KafkaClusterTest.java

```diff
@@ -47,140 +59,134 @@ public class KafkaClusterTest {
+        // TODO: waiting for new versions support in ozangunalp repo https://github.com/ozangunalp/kafka-native/issues/21
+        // version("3.3.1"),
+        // version("3.2.1")
+        );
     }
 
-    @Test
-    public void kafkaClusterZookeeperMode() throws Exception {
-        try (var cluster = KafkaClusterFactory.create(KafkaClusterConfig.builder()
-                .testInfo(testInfo)
-                .kraftMode(false)
-                .build())) {
-            cluster.start();
-            verifyRecordRoundTrip(1, cluster);
-        }
+    private static Stream<BrokerCluster> clusterSizes() {
+        return Stream.of(
+                brokerCluster(1),
+                brokerCluster(2));
     }
 
-    @Test
-    public void kafkaTwoNodeClusterKraftMode() throws Exception {
-        int brokersNum = 2;
-        try (var cluster = KafkaClusterFactory.create(KafkaClusterConfig.builder()
-                .testInfo(testInfo)
-                .brokersNum(brokersNum)
-                .kraftMode(true)
-                .build())) {
-            assumeTrue(cluster instanceof TestcontainersKafkaCluster, "KAFKA-14287: kraft timing out on shutdown in multinode case");
-            cluster.start();
-            verifyRecordRoundTrip(brokersNum, cluster);
-        }
+    private final Object[] brokerClusters = clusterSizes().toArray();
+
+    static Stream<List<Annotation>> clusterSizesAndModes() {
+        return Stream.of(
+                List.of(brokerCluster(1), kraftCluster(1)),
+                List.of(brokerCluster(2), kraftCluster(1)),
+                List.of(brokerCluster(1), zooKeeperCluster()),
+                List.of(brokerCluster(2), zooKeeperCluster()));
     }
 
-    @Test
-    public void kafkaTwoNodeClusterZookeeperMode() throws Exception {
-        int brokersNum = 2;
-        try (var cluster = KafkaClusterFactory.create(KafkaClusterConfig.builder()
-                .testInfo(testInfo)
-                .brokersNum(brokersNum)
-                .kraftMode(false)
-                .build())) {
-            cluster.start();
-            verifyRecordRoundTrip(brokersNum, cluster);
+    @BeforeEach
+    public void initializeIndexes() {
+        if (clusterSizesIndex == clusterSizes().count()) {
+            clusterSizesIndex = 0;
         }
     }
 
-    @Test
-    public void kafkaClusterKraftModeWithAuth() throws Exception {
-        try (var cluster = KafkaClusterFactory.create(KafkaClusterConfig.builder()
-                .kraftMode(true)
-                .testInfo(testInfo)
-                .securityProtocol("SASL_PLAINTEXT")
-                .saslMechanism("PLAIN")
-                .user("guest", "guest")
-                .build())) {
-            cluster.start();
-            verifyRecordRoundTrip(1, cluster);
-        }
+    @TestTemplate
+    @ExtendWith(KafkaClusterExtension.class)
+    public void kafkaClusterTemplate(@ConstraintsMethodSource("clusterSizesAndModes") InVMKafkaCluster cluster)
```

Comment:
Does the parameter type need to be the concrete `InVMKafkaCluster`? I would expect it to be the `KafkaCluster` interface. 

```suggestion
    public void kafkaClusterTemplate(@ConstraintsMethodSource("clusterSizesAndModes") KafkaCluster cluster)
```

Reply:
As `@Version` does not make sense for `IN_VM` tests, the only way I found to force to test in a container is using `TestcontainerKafkaCluster` instead of `KafkaCluster`. So, just in case, the other tests shall be `InVMKafkaCluster` to avoid any possible duplication

Reply:
I'm confused why doesn't `@Version` make sense in `IN_VM`? I can appreciate it being harder to orchestrate a specific version `IN_VM` but to me the requirement still makes sense. 

Also if we have tests which only make sense for one provisioning strategy I'd prefer them to be in their own file and explicitly configure the strategy, but that could just be me.

Reply:
AFAIK the kafka version for `IN_VM` comes from pom file, so it cannot be set dynamically inside the test, that's why I said that it does not make sense for that kind of tests.

We could use different files, it is not a problem from my side. WDYT @tombentley @k-wall ?

Reply:
Reworked